clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusClusterProvider.groovy

```diff
@@ -195,6 +215,20 @@ class TitusClusterProvider implements ClusterProvider<TitusCluster>, ServerGroup
     serverGroup
   }
 
+  private Optional<Map<String, String>> getTaggedImageAttributes(String account, Job job) {
+    try {
+      // Infer registry from account configuration
+      String registry = TitusUtils.getRegistry(accountCredentialsProvider, account)
+      String key = DockerRegistryCacheKeys.getTaggedImageKey(registry, job.applicationName, job.version)
+      Set<CacheData> images = DockerRegistryProviderUtils
+        .getAllMatchingKeyPattern(cacheView, DockerRegistryCacheKeys.Namespace.TAGGED_IMAGE.getNs(), key)
+      return images.collect { it.attributes }
+    }
+    catch (Exception e) {
+      log.error("Failed to fetch tagged image attributes ", e)
+    }
+  }
```

Comment:
Continuing from: https://github.com/spinnaker/clouddriver/pull/5035#discussion_r509855514

The method signature was changed to `Optional<Map<String, String>>` but the method does not do anything to make the method return an optional, and the one place that it returns a value, it's of type `List<Map<String, String>>`. I haven't tested this code that I'm about to write, but I would have expected to see something more like this:

```
  private Optional<Map<String, String>> getTaggedImageAttributes(String account, Job job) {
    try {
      // Infer registry from account configuration
      String registry = TitusUtils.getRegistry(accountCredentialsProvider, account)
      String key = DockerRegistryCacheKeys.getTaggedImageKey(registry, job.applicationName, job.version)
      Set<CacheData> images = DockerRegistryProviderUtils
        .getAllMatchingKeyPattern(cacheView, DockerRegistryCacheKeys.Namespace.TAGGED_IMAGE.getNs(), key)
      List<Map<String, String>> allAttributes = images.collect { it.attributes }
      if (allAttributes.size() > 1) {
        throw new TItusException("Expected 1 tagged image while retrieving attributes, but got ${allAttributes.size()}")
      }
      return Optional.of(allAttributes.first())
    }
    catch (Exception e) {
      log.error("Failed to fetch tagged image attributes ", e)
    }
    return Optional.empty()
  }
```

The key changes being:

1. The end of the method body returns `Optional.empty()` if we are in an exceptional case, instead of implicitly returning `null`.
2. Validate that we're indeed only getting one image from `getAllMatchingKeyPattern`. Since this method returns a `Set`, I would assume that getting more than 1 image is an error condition, since `Set`s are unordered and thus this method becomes non-deterministic.
3. Returns `Optional.of(allAttributes.first())` if we indeed get only one image from the query.

I ask if you tested this because running this method as it exists today would cause a class cast exception in the case where you're returning a result, or an NPE in the case where no image was found.

Reply:
Makes sense what you mention here , from testing it locally it is returning an Optional , i'll validate a few more cases and update the code based on your inputs here.